agent/lib/kontena/workers/service_pod_worker.rb

```diff
@@ -0,0 +1,106 @@
+require_relative '../service_pods/creator'
+require_relative '../service_pods/stopper'
+require_relative '../service_pods/terminator'
+
+module Kontena::Workers
+  class ServicePodWorker
+    include Celluloid
+    include Kontena::Logging
+    include Kontena::ServicePods::Common
+    include Kontena::Helpers::RpcHelper
+
+    attr_reader :node, :prev_state, :service_pod
+    attr_accessor :service_pod
+
+    def initialize(node, service_pod = nil)
+      @node = node
+      @service_pod = service_pod
+      @prev_state = 'unknown'
+    end
+
+    def update(service_pod = nil)
+      @service_pod = service_pod if service_pod
+      ensure_desired_state
+    end
+
+    def destroy
+      @service_pod.mark_as_terminated
+      update
+    end
+
+    def ensure_desired_state
+      return unless service_pod
+
+      exclusive {
+        debug "state of #{service_pod.name}: #{service_pod.desired_state}"
+        service_container = get_container(service_pod.service_id, service_pod.instance_number)
+        if service_pod.running? && service_container.nil?
+          info "creating #{service_pod.name}"
+          ensure_running
+        elsif service_container && service_pod.running? && !service_container.running?
+          info "starting #{service_pod.name}"
+          ensure_running
+        elsif service_pod.running? && (service_container && service_container.service_revision != service_pod.service_revision)
```

Comment:
Some design thoughts re the currently missing support for restarting the container here: The `kontena service restart ...` should set some `restart_at` timestamp on the `GridService` / `GridServiceInstance`s, and then this agent `ServicePodWorker` can trigger a restart if `service_container.state.started_at < service_pod.restart_at`.